src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+import type { ModelOptions } from './types/model.js';
+import { createAblyApp } from './utilities/test/createAblyApp.js';
+import { getEventPromises } from './utilities/test/promises.js';
+
+interface TestContext extends ModelOptions {
+  model: Model<any>;
+  channelName: string;
+  eventData: {
+    name: string;
+    data: any;
+    mutationId: string;
+  };
+}
+
+describe('Stream integration', () => {
+  beforeEach<TestContext>(async (context) => {
+    const channelName = 'test-channel';
+    const data = await createAblyApp({
+      keys: [{}],
+      namespaces: [{ id: channelName, persisted: true }],
+      channels: [
+        {
+          name: channelName,
+          presence: [
+            { clientId: 'John', data: 'john@test.com' },
+            { clientId: 'Dave', data: 'dave@test.com' },
+          ],
+        },
+      ],
+    });
+    const ably = new Realtime({
+      key: data.keys[0].keyStr,
+      environment: 'sandbox',
+    });
+    const logger = pino({ level: 'debug' });
+    const model = new Model(
+      channelName,
+      {
+        sync: async (data) => data,
+        merge: (state, event) => (state ? { ...state, ...event.data } : event.data),
+      },
+      {
+        ably,
+        channelName,
+        logger,
+        syncOptions: defaultSyncOptions,
+        optimisticEventOptions: defaultOptimisticEventOptions,
+        eventBufferOptions: defaultEventBufferOptions,
+      },
+    );
+
+    context.model = model;
+    context.channelName = channelName;
+    context.ably = ably;
+
+    context.eventData = {
+      name: 'update',
+      data: {
+        noOfMessages: 34,
+      },
+      mutationId: 'some-id-1',
+    };
+  });
+
+  afterEach<TestContext>(async ({ model }) => {
+    await model.dispose();
+    vi.restoreAllMocks();
+  });
+
+  it<TestContext>('changes state on sync, pause, resume, dispose', async ({ model }) => {
+    expect(model.state).toEqual('initialized');
+
+    const sync = model.sync({
+      data: {
+        foo: 'foobar',
+        bar: {
+          baz: 1,
+        },
+      },
+      sequenceId: '1',
+    });
+    expect(model.state).toEqual('syncing');
+
+    await sync;
+    expect(model.state).toEqual('ready');
+
+    await model.pause();
+    expect(model.state).toEqual('paused');
+
+    await model.resume();
+    expect(model.state).toEqual('ready');
+
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
```

Comment:
Here we are setting the optimistic data to the same data the model was intialised with from the `sync` call. I think it would be more robust to apply different data optimistically in this test (and in the others) and assert on that